wordpress-mobile/WordPress-iOS
# Send highest priority plan in Zendesk Tickets

Fixes #14224

Related WordPressKit PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/250

Previously, if users had more than one site/plan, Zendesk could assign the wrong plan.
What we want, is to always set the 'highest' plan in the ticket, regardless of the site.

The priority should be:
1. eCommerce
2. Business/Professional
3. Premium
4. Personal
5. Blogger
6. Free

More context can be found here:
- p8wKgj-1eQ-p2
- p8wKgj-1gx-p2

#### Notes
1. Uses the /plans/mobile endpoint to retrieve the plan priority, in order to be 'future proof' if the list of available plans changes

2. I am now populating `sitePlanCache` even in the case of English language users. This is because, if a user adds/removes a plan, the `allBlogs` collection might not be updated (it gets updated only with login/logout or visiting `BlogListViewController`, that contains the `syncBlogs()` method), so for these users, the list might not be updated just yet, when they request a ticket. It's likely an edge case, but we are still doing that request for non English users anyway

3. In the event that there are unrecognized plans (new plans that have been added before we could update the app), these will be sent to Zendesk as tags in the ticket, in the form of
  "unrecognized_plan_[plan_name]"

To test:

Prerequisites:
- a Zendesk account
- Another WordPress.com account, different from the previous one, that has a variety of plans (or just add remove plans on the store admin)

1. Log in with the second account
2. Open one or more requests on Zendesk and verify that the assigned plan priority list matches the above one.
3. Optionally, hardcode a list of random strings in `plans` in the method `getHighestPriorityPlans()` of `ZenDeskUtils` and verify that the unrecognized plans are sent to Zendesk as ticket tags.

PR submission checklist:

- [x] I have considered adding unit tests where possible. - Added
- [x] ~~I have considered adding accessibility improvements for my changes.~~ No impact on the UI
- [x] ~~I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.~~ No user facing changes
